File: 삼성알고리즘특강/3일차/집배원한상덕/2842.py
from collections import deque
dy = [ -1, 1, 0, 0 , -1, -1, 1 ,1]
dx = [0,0,-1,1,-1,1,-1,1]
y_post, x_post = 0, 0
vil = []
h = []
cnt_k = 0
# 데이터 입력 받기
n = int(input())
for _ in range(n):
   vil.append(list(input()))

for i in range(n):
    for j in range(n):
        if vil[i][j] == 'K':
            cnt_k += 1
        elif vil[i][j] == 'P':
            y_post = i
            x_post = j

for _ in range(n):
    h.append(list(map(int,input().split())))
hhh = sum(h, [])

def bfs(low, high):
    cnt = 0
    vt = [[False] *50 for _ in range(50)]
    q = deque()
    q.append([y_post, x_post])
    vt[y_post][x_post] = True

    if (h[y_post][x_post] < low or high < h[y_post][x_post]):
        return -1

    while( q and cnt < cnt_k):
        cur = q.popleft()
        for i in range(8):
            nxty = cur[0] + dy[i]
            nxtx = cur[1] + dx[i]
            if nxty < 0 or n <= nxty or nxtx < 0 or n <= nxtx:
                continue
            if vt[nxty][nxtx]:
                continue
            if h[nxty][nxtx] < low or high < h[nxty][nxtx]:
                continue
            if vil[nxty][nxtx] == 'K':
                cnt += 1

            vt[nxty][nxtx] = True
            q.append([nxty,nxtx])
    return cnt

hhh.sort()
answer = hhh[-1] - hhh[0]
low, high = 0, 0
while (low< len(hhh) and high < len(hhh) and low <= high):
    if bfs(hhh[low], hhh[high]) == cnt_k:
        tmp = hhh[high] -hhh[low]
        if tmp < answer:
            answer = tmp
        low += 1
    else:
        high += 1

print(answer)

File: 삼성알고리즘특강/3일차/집배원한상덕/2842.cpp
#include <cstdio>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

typedef pair<int, int> pii;

int n, h[50][50], dy[] = { -1, 1, 0, 0, -1, -1, 1, 1 }, dx[] = { 0, 0, -1, 1, -1, 1, -1, 1 };
int cnt_k, y_post, x_post;
char vil[50][51];
vector<int> hhh;

// low ~ high의 고도사이로 탐색을 했을때, 배달 할 수 있는 집의 개수를 리턴
int bfs(int low, int high) {
	// to-do
	int cnt = 0;
	bool vt[50][50] = { false, };
	queue<pii> q;
	q.push(pii(y_post, x_post));
	vt[y_post][x_post] = true;
	if (h[y_post][x_post] < low || high < h[y_post][x_post]) return -1;
	while (!q.empty() && cnt < cnt_k) {
		pii cur = q.front();
		q.pop();
		for (int i = 0; i < 8; i++) {
			int nxty = cur.first + dy[i], nxtx = cur.second + dx[i];
			if (nxty < 0 || n <= nxty || nxtx < 0 || n <= nxtx) continue;
			if (vt[nxty][nxtx]) continue;
			if (h[nxty][nxtx] < low || high < h[nxty][nxtx]) continue;
			if (vil[nxty][nxtx] == 'K') {
				cnt++;
			}
			vt[nxty][nxtx] = true;
			q.push(pii(nxty, nxtx));
		}
	}

	return cnt;
}

// low ~ high의 고도사이로 탐색을 했을때, 배달 할 수 있는 집의 개수를 리턴

//int dfs(int low, int high) {
//  // to-do
//}

int main() {
	//freopen("res/B2842.in", "r", stdin);
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		scanf("%s", vil[i]);
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (vil[i][j] == 'K') {
				cnt_k++;
			}
			else if (vil[i][j] == 'P') {
				y_post = i, x_post = j;
			}
		}
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			scanf("%d", &h[i][j]);
			hhh.push_back(h[i][j]);
			printf("%d", h[i][j]);
		}
	}
}